src/txmempool.cpp

```diff
@@ -186,10 +200,13 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
         staged_ancestors.erase(stage);
         totalSizeWithAncestors += stageit->GetTxSize();
 
-        if (stageit->GetSizeWithDescendants() + entry.GetTxSize() > limitDescendantSize) {
+        // When multiple transactions are passed in, the ancestors and descendants of all transactions
+        // considered together must be within limits even if they are not interdependent. This may be
+        // stricter than the limits for each individual transaction.
```

Comment:
I agree -- it should be one connected component. The only "quirk" of this is that "packages" can get "invalidated" after a block is mined.

Reply:
Distinct `testmempoolaccept` calls returning different results for policy reasons is not a "quirk," it's just a natural part of mempool policy. With single test accepts today you could call it 3 times and get 3 different results for the same tx because anything can happen in between calls.

I agree that p2p packages shouldn't have independent transactions, but we're not really there yet. This PR is just adding dry-run logic reachable through RPC, and we'll write docs to make it clear what this should be used for (i.e. not for batch validation). Obviously I wouldn't want to create a footgun for users who misunderstand/don't read the release notes (which is why I set a limit of 25 for example), but I don't think that's the case here.

TLDR: I agree conceptually, but don't think it's necessary in this PR. I'm trying to keep this as bare-bones as possible and it's 1000+ lines...

Reply:
> Distinct testmempoolaccept calls returning different results for policy reasons is not a "quirk," it's just a natural part of mempool policy. With single test accepts today you could call it 3 times and get 3 different results for the same tx because anything can happen in between calls.

And I don't agree with this assumption of anything can happen in between calls. Ultimately you can isolate and disconnect your node to verify the correctness equivalence of `testmempoolaccept` of package A of size N compare to N sequential call to `testmempoolaccept` for each member of package A.

That said what previous discussions about RBF conflict sounds to have establish is package policy checks != mempool policy checks, so I don't think we need to agree on assumptions there. We'll make it clear to the user that you can succeed a package A but individually fail transaction member or vice-versa.

> TLDR: I agree conceptually, but don't think it's necessary in this PR. I'm trying to keep this as bare-bones as possible and it's 1000+ lines...

IIRC, it was on my push to add ancestors/descendants at the package-level but given the complexity in practice I would prefer not to touch `txmempool.cpp` in this PR. I think we can roll-back e7aa139 and introduce in its own PR ? That would make more happy reviewer to not have think about what we might breack by modifying `CalculateMemPoolAncestors` :)

Reply:
> I think we can roll-back e7aa139 and introduce in its own PR ? That would make more happy reviewer to not have think about what we might breack by modifying CalculateMemPoolAncestors :)

@ariard Yeah, good point! Without the ancestor/descendant logic I think we could end up with a chain of 49 at least, but this is just test accepts so it won't really hurt us... I've just reordered the commits so everything past the release notes could go in a followup PR, maybe that'd help?

Reply:
separated to #21800